ev3dev2/led.py

```diff
@@ -346,3 +354,214 @@ def all_off(self):
+
+            even = not even
+            sleep(self.sleeptime)
+
+
+class LedCycle(LedAnimate):
+    """
+    Cycle all LEDs through ``colors``. Do this on a loop where
+    we display each color for ``sleeptime`` seconds.
+
+    Example:
+
+    .. code:: python
+        from ev3dev2.led import LedCycle
+        from time import sleep
+        leds = LedCycle(('AMBER', 'RED', 'GREEN'))
+
+        leds.start()
+        sleep(10)
+        leds.join()
+    """
+
+    def __init__(self, colors, groups=('LEFT', 'RIGHT'), sleeptime=0.5):
+        super(LedCycle, self).__init__()
+        self.colors = colors
+        self.groups = groups
+        self.sleeptime = sleeptime
+
+    def run(self):
+        index = 0
+        max_index = len(self.colors)
+
+        while not self.done.isSet():
+            for group in self.groups:
+                self.leds.set_color(group, self.colors[index])
+            index += 1
+
+            if index == max_index:
+                index = 0
+
+            sleep(self.sleeptime)
+
+
+class LedRainbow(LedAnimate):
+    """
+    Gradually fade from one color to the next
+
+    Example:
+    .. code:: python
+        from ev3dev2.led import LedRainbow
+        from time import sleep
+        leds = LedRainbow()
+
+        leds.start()
+        sleep(10)
+        leds.join()
+    """
+
+    def __init__(self, group1='LEFT', group2='RIGHT', increment_by=0.1, sleeptime=0.1):
+        super(LedRainbow, self).__init__()
+        self.group1 = group1
+        self.group2 = group2
+        self.increment_by = increment_by
+        self.sleeptime = sleeptime
+
+    def run(self):
+        """
+        state 0: (LEFT,RIGHT) from (0,0) to (1,0)...RED
+        state 1: (LEFT,RIGHT) from (1,0) to (1,1)...AMBER
+        state 2: (LEFT,RIGHT) from (1,1) to (0,1)...GREEN
+        state 3: (LEFT,RIGHT) from (0,1) to (0,0)...OFF
+        """
+        state = 0
+        left_value = 0
+        right_value = 0
+        MIN_VALUE = 0
+        MAX_VALUE = 1
+        self.leds.all_off()
+
+        while not self.done.isSet():
+
+            if state == 0:
+                left_value += self.increment_by
+            elif state == 1:
+                right_value += self.increment_by
+            elif state == 2:
+                left_value -= self.increment_by
+            elif state == 3:
+                right_value -= self.increment_by
+            else:
+                raise Exception("Invalid state {}".format(state))
+
+            # Keep left_value and right_value within the MIN/MAX values
+            left_value = min(left_value, MAX_VALUE)
+            right_value = min(right_value, MAX_VALUE)
+            left_value = max(left_value, MIN_VALUE)
+            right_value = max(right_value, MIN_VALUE)
+
+            self.leds.set_color(self.group1, (left_value, right_value))
+            self.leds.set_color(self.group2, (left_value, right_value))
+
+            if state == 0 and left_value == MAX_VALUE:
```

Comment:
Not following you?

Reply:
ok for me that is much much harder to follow than a basic state machine :)  With w=4 (so that the max is 1 like we need) this ends up cycling through the following
```
2018-11-24 13:35:31,396  INFO: left_value 1.0, right_value 0.0
2018-11-24 13:35:31,552  INFO: left_value 1.0, right_value 1.0
2018-11-24 13:35:31,684  INFO: left_value 0.0, right_value 1.0
2018-11-24 13:35:31,812  INFO: left_value 1.0, right_value 1.0
2018-11-24 13:35:31,939  INFO: left_value 1.0, right_value 0.0
```

Reply:
If you'd prefer to use the state machine approach and it works for the various platforms, feel free to go ahead with that — whatever works is fine with me as long as it isn't super ugly.

Reply:
👍

Reply:
I should have some time to finish this one up over the holidays